File: movies/movies.R
users<- read.table('users.txt',sep =":",encoding = 'utf-8',colClasses = c(NA,"NULL")) #The text file importing functions only support single characters as column separators. However, you can tell read.table to ignore columns for import with its colClasses parameter (see the help file):
colnames(users)<-c('user_id','gender','age','occupation','zip')
head(users)

ratings <- read.csv('ratings.txt',sep=":",encoding = 'utf-8',colClasses = c(NA,"NULL"),fill = T)
colnames(ratings) <- c('user_id','movie_id','rating','timestap')
head(ratings)

movies <- read.csv('movies.txt',sep = ":",encoding = 'utf-8',colClasses = c(NA,"NULL"),fill = T)
colnames(movies) <- c('movie_id','title','genres')

movie_data<-merge(users,ratings,by.x = 'user_id',by.y = 'user_id')
movie_data<-merge(movie_data,movies,by.x = 'movie_id',by.y = 'movie_id')
head(movie_data)
movie_data<-data.frame(movie_data)

library(dplyr)
active_movie<-movie_data %>% group_by(movie_id) %>% dplyr::summarise(count=n()) %>% filter(count>250)
mean_rating<- movie_data %>% group_by(movie_id,gender)%>%dplyr::summarise(mean_rating=mean(rating),count=n()) %>% filter(movie_id  %in% active_movie$movie_id)
head(mean_rating)

#---------------------------------------reshape-------------------------------------------------------
library(reshape2)
#wide to long
head(mtcars)
mt<-mtcars
mt['car_type']<-row.names(mtcars)
rownames(mt)<-NULL
head(mt)
melt(mt,id.vars = 'car_type',variable.name = 'item_type',value.name = 'value')
melt(mt,id.vars = c('car_type','mpg'),variable.name = 'item_type',value.name = 'value') #id.vars = c('car_type','mpg')
m_mt<-melt(mt,id.vars = c('car_type','mpg')) #id.vars = c('car_type','mpg')
acast(m_mt,carb~cyl~variable,mean)

head(airquality)
names(airquality) <- tolower(names(airquality))
aqm <- melt(airquality, id=c("month", "day"), na.rm=TRUE)
head(aqm)

acast(aqm, day ~ month ~ variable)
acast(aqm, month ~ variable, mean)
acast(aqm, month ~ variable, mean, margins = TRUE) # margins = TRUE 计算边际
dcast(aqm, month ~ variable, mean, margins = c("month", "variable"))

library(plyr) # needed to access . function
acast(aqm, variable ~ month, mean, subset = .(variable == "ozone")) #.(variable==) 筛选变量
acast(aqm, variable ~ month, mean, subset = .(month == 5))

#Chick weight example
head(ChickWeight)
names(ChickWeight) <- tolower(names(ChickWeight))
chick_m <- melt(ChickWeight, id=2:4, na.rm=TRUE)
head(chick_m)

dcast(chick_m, time ~ variable, mean) # average effect of time
dcast(chick_m, diet ~ variable, mean) # average effect of diet
acast(chick_m, diet ~ time, mean) # average effect of diet & time

# How many chicks at each time? - checking for balance
acast(chick_m, time ~ diet, length)
acast(chick_m, chick ~ time, mean)
acast(chick_m, chick ~ time, mean, subset = .(time < 10 & chick < 20))

acast(chick_m, time ~ diet, length)

head(chick_m)
dcast(chick_m, diet + chick ~ time)
acast(chick_m, diet + chick ~ time)
acast(chick_m, chick ~ time ~ diet)
acast(chick_m, diet + chick ~ time, length, margins="diet")
acast(chick_m, diet + chick ~ time, length, drop = FALSE)

#Tips example
dcast(melt(tips), sex ~ smoker, mean, subset = .(variable == "total_bill"))

ff_d <- melt(french_fries, id=1:4, na.rm=TRUE)
acast(ff_d, subject ~ time, length)
acast(ff_d, subject ~ time, length, fill=0)
dcast(ff_d, treatment ~ variable, mean, margins = TRUE)
dcast(ff_d, treatment + subject ~ variable, mean, margins="treatment")
if (require("lattice")) {
  lattice::xyplot(`1` ~ `2` | variable, dcast(ff_d, ... ~ rep), aspect="iso")
}


#---------------------------计算男女评分分歧------------------------------------
mean_rating1<-mean_rating[,1:3]
head(mean_rating1)
#先melt
re_mean_rating1<-melt(mean_rating1,id=c('movie_id','gender'))
head(re_mean_rating1)
#再cast
(a<-acast(re_mean_rating1,movie_id~gender))
a
a<-data.frame(a)
a$movie_id<-row.names(a)
head(a)


a<-a%>%mutate(diff=abs(F-M)) # 等价 a['diff']<-abs(a['F']-a['M'])
head(a)
a<-arrange(a,desc(diff))
head(a)
a<-a[,c(3,1,2,4)] # 更改列的顺序
head(a)


#-------------------------计算评分波动性-----------------------------
rating_var<-movie_data %>% filter(movie_id %in% active_movie$movie_id) %>% group_by(movie_id) %>% dplyr::summarise(std=var(rating)) %>% arrange(desc(std))
head(rating_var)

File: movies/movies.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jun 12 20:00:29 2017

@author: ranjing
"""

import requests
import pandas as pd
import numpy as np
#下载保存数据
movies=requests.get("https://raw.githubusercontent.com/wesm/pydata-book/master/ch02/movielens/movies.dat")
ratings=requests.get("https://raw.githubusercontent.com/wesm/pydata-book/master/ch02/movielens/ratings.dat")
users=requests.get("https://raw.githubusercontent.com/wesm/pydata-book/master/ch02/movielens/users.dat")    
with open('./movies.txt','w',encoding='utf-8') as f:
    f.write(movies.text)
with open('./ratings.txt','w',encoding='utf-8') as f:    
    f.write(ratings.text)
with open('./users.txt','w',encoding='utf-8') as f:
    f.write(users.text)

#读取数据
unames=['user_id','gender','age','occupation','zip']
users=pd.read_table('users.txt',sep="::",header=None,names=unames)    

rnames=['user_id','movie_id','rating','timestap']
ratings=pd.read_table('ratings.txt',sep="::",header=None,names=rnames)

mnames=['movie_id','title','genres']
movies=pd.read_table('movies.txt',sep="::",header=None,names=mnames)

users[:5]
ratings[:5]
movies[:5]

#汇集数据
data=pd.merge(pd.merge(ratings,users,on='user_id'),movies,on='movie_id')
##保存数据
with open('movie_data.txt','w',encoding='utf-8') as f:
    f.write(data)
data[0:1] #查看第一行
data.ix[0]# 查看第一行

##分类汇总 picot_table(透视表)
mean_ratings=data.pivot_table('rating',index='title',columns='gender',aggfunc=[np.mean,len],fill_value=0)
mean_ratings[:5]
mean_ratings[('mean','F')] # 层次化索引

''' 和上面作用一样
mean_ratings=pd.pivot_table(data,index='title',columns='gender',values='rating',aggfunc=[np.mean,len])
mean_ratings[:5]

mean_ratings=pd.pivot_table(data,index=['title','genres','gender'],values='rating',aggfunc=[np.mean])
mean_ratings[:50]
'''

##active titles
ratings_by_title =data.groupby('title').size()
active_titles=ratings_by_title[ratings_by_title>250].index
mean_ratings=mean_ratings.ix[active_titles]

''' 和上面的作用一样
ratings_by_title=data.pivot_table(index='title',values='genres',aggfunc=len)
active_titles=ratings_by_title[ratings_by_title>250].index
mean_ratings.ix[active_titles]
'''

##排序
mean_ratings.sort_index(by=('mean','F'),ascending=False) # 层次化索引 by=('mean','F')

##计算男女评分分歧
mean_ratings['diff']=np.abs(mean_ratings[('mean','F')]-mean_ratings[('mean','M')]) # 增加一行
mean_ratings_sort_by_diff=mean_ratings.sort_index(by='diff',ascending=False)

##计算评分波动
rating_std=data.groupby('title')['rating'].std()[active_titles]
#rating_std=pd.DataFrame(rating_std,columns=['std']) # 没有加[]的时候，显示错误Index(...) must be called with a collection of some kind, 'std' was passed
rating_std[:5]
rating_std.order(ascending=False)[:10]




























